volpiano_display_utilities/text_volpiano_alignment.py

```diff
@@ -6,238 +6,317 @@
 
 Use align_text_and_volpiano to align the text and melody strings of a chant.
 """
-import re
 import logging
 from itertools import zip_longest
-from typing import List, Tuple
+from typing import List, Tuple, TypeVar
 from .cantus_text_syllabification import syllabify_text
-
-# A string containing all valid volpiano characters in Cantus
-# Database. Used to check for invalid characters.
-INVALID_VOLPIANO_CHARS_REGEX = re.compile(
-    r"[^\-19abcdefghijklmnopqrsyz)ABCDEFGHIJKLMNOPQRSYZ3467]"
+from .syllabified_section import SyllabifiedTextSection, SyllabifiedVolpianoSection
+from .volpiano_syllabification import (
+    syllabify_volpiano,
+    preprocess_volpiano,
+    adjust_music_spacing,
+    adjust_missing_music_spacing,
 )
 
-# Matches any material before the clef, the clef itself, and
-# any following spacing in a volpiano string.
-STARTING_MATERIAL_REGEX = re.compile(r".*1-*")
-
-# Split sections of volpiano on clefs, barline markers, and missing
-# music indicators. Includes spacing (hyphens) and section markers (7's)
-# in the split, if present.
-VOLPIANO_SECTIONING_REGEX = re.compile(r"([134][\-7]*|6[\-7]*6[\-7]*)")
-
-VOLPIANO_WORD_REGEX = re.compile(r".*?---")
-
-VOLPIANO_SYLLABLE_REGEX = re.compile(r".*?-{2,3}")
+T = TypeVar("T", str, List, SyllabifiedTextSection, SyllabifiedVolpianoSection)
 
 
-def _preprocess_volpiano(raw_volpiano_str: str) -> str:
+def _zip_and_align(
+    text: List[T], volpiano: List[T], pad_text: T, pad_volpiano: T
+) -> List[Tuple[T, T]]:
```

Comment:
It took me a while to figure out what `T` is doing here. I gather that "T" is the convention, but could it be renamed to something more descriptive - `Section` or `SectionType` (not sure what the best approach is here - still trying to wrap my head around `TypeVar`s)?

Thinking bigger-picture, why is it necessary that `_zip_and_align` be able to handle all of `str`, `list`, and `Syllabified*Section`s? Are you using all of this functionality?

If this were to handle only `str`s, say, then you could use `pad_text: str = " ", pad_volpiano: str = "-"`, couldn't you, which is how the function will almost always be used? And even if you continue to use `T` to allow various types for `text` and `volpiano`, `pad_volpiano` will always be a `str` and never a `SyllabifiedTextSection`, won't it?

Reply:
So using a TypeVar ensures that all the arguments labelled with `SyllableOrWordT` will be of the same type? If this is the case, I gather it will prevent situations where `text` is a list of `str`s and `volpiano` is a list of `list[str]`s, which is what I was worried about.

Reply:
Looking through the places where `_zip_and_align` is used in `_align_word` and `_align_section`, I see why there's the need for `_zip_and_align` to accept both types as argument. Thanks for taking the time to address the other parts of the TypeVar.